# HG changeset patch
# Parent 253d2b1af470f94b7d42674827729eae31ac5420
CA-49885: Restrict Intel Nehalem and Westmere class processors to a maximum 
c-state level of 1 to workaround various Intel errata assoicated with these 
processors.

diff -r 253d2b1af470 xen/arch/x86/acpi/cpu_idle.c
--- a/xen/arch/x86/acpi/cpu_idle.c
+++ b/xen/arch/x86/acpi/cpu_idle.c
@@ -856,6 +856,41 @@ static int check_cx(struct acpi_processo
 static unsigned int latency_factor = 2;
 integer_param("idle_latency_factor", latency_factor);
 
+/*
+ * Detect Intel Nehalem CPU's with C-states enabled and restrict to C1 due to
+ * Intel Errata BA80, AAK120, AAM108, AAO67, BD59, AAY54
+ */
+void nehalem_cpu_cstate_workaround(struct acpi_processor_cx *cx)
+{
+    static bool_t nehalem_cstate_errata_workaround_warning = 0;
+    struct cpuinfo_x86 *c = &current_cpu_data;
+
+    if ( cx->entry_method == ACPI_CSTATE_EM_NONE )
+	    return;
+    if ( c->x86_vendor == X86_VENDOR_INTEL && c->x86 == 6 )
+    {
+        switch ( c->x86_model )
+        {
+        /* Nehalem */
+        case 0x1A:
+        case 0x1E:
+        case 0x1F:
+        case 0x2E:
+        /* Westmere */
+        case 0x25:
+        case 0x2C:
+            if ( !nehalem_cstate_errata_workaround_warning )
+            {
+                printk(XENLOG_WARNING "Disabling C-states C3 and C6 on Nehalem"
+                        " Processors due to errata\n");
+                nehalem_cstate_errata_workaround_warning = 1;
+            }
+            max_cstate = 1;
+            break;
+        }
+    }
+}
+
 static void set_cx(
     struct acpi_processor_power *acpi_power,
     xen_processor_cx_t *xen_cx)
@@ -905,6 +940,8 @@ static void set_cx(
         cx->entry_method = ACPI_CSTATE_EM_NONE;
     }
 
+    nehalem_cpu_cstate_workaround(cx);
+
     cx->latency  = xen_cx->latency;
     cx->power    = xen_cx->power;
     
